REST V2: Key ranges for entries + diff

snazy opened the pull request:
Builds upon the work in #5656.

However, the "get range of contents" has been removed, because it's tricky to add it to the current Nessie Java API and since it's probably not a very useful functionality. It would also require adding paging and min/max-keys to the `GetMultipleContentRequest`/`Response` types to leverage the existing paging infrastructure and/or add `ContentsParams` parameter object - neither works well with the existing code.

Adds the necessary `VersionStore` changes, which are pretty straight forward.

Fixes #3957
Fixes #5591

dimas-b commented on api/model/src/main/java/org/projectnessie/api/v2/params/KeyRangeParams.java:
```diff
@@ -0,0 +1,69 @@
+  @QueryParam("max-key")
+  @Nullable
+  @jakarta.annotation.Nullable
+  private ContentKey maxKey;
```
Just wondering: Would it make sense to have a separate `key-prefix` query parameter? We could validate it to actually be the prefix of both min and max keys if all are set.

dimas-b commented on versioned/storage/store/src/main/java/org/projectnessie/versioned/storage/versionstore/VersionStoreImpl.java:
```diff
@@ -494,8 +496,8 @@ public PaginationIterator<KeyEntry> getKeys(Ref ref, String pagingToken, boolean
-            : null;
-    StoreKey end = null;
+            : (minKey != null ? keyToStoreKey(minKey) : null);
+    StoreKey end = maxKey != null ? keyToStoreKey(maxKey) : null;
```
Will this work when paginating over a "prefix" query? I mean that on the second page min and max keys will no longer be equals when passed to `index.iterator(...)` :thinking:

snazy replied:
The max key will still be the same, and "min" would be "replaced" with the key from the paging token.

dimas-b replied:
that's exactly what I mean. If we start with min == max for a prefix query, min != max in the second page... Just wondering if we want this to work for prefix queries as well as for normal ranges..

dimas-b replied:
Sorry, my comment is not relevant to the latest code - resolving.

dimas-b commented on api/client/src/main/java/org/projectnessie/client/http/v2api/HttpGetDiff.java:
```diff
@@ -36,20 +38,33 @@ protected DiffParams params() {
+    keys.forEach(k -> req.queryParam("key", k.toPathString()));
+    ContentKey k = params.minKey();
+    if (k != null) {
+      req.queryParam("min-key", k.toPathString());
```
Should we have this in `HttpGetEntries` too?

snazy commented on api/model/src/main/java/org/projectnessie/api/v2/doc/ApiDoc.java:
```diff
@@ -109,12 +109,20 @@ public interface ApiDoc {
+          + KEY_ELEMENTS_DESCRIPTION;
+
+  String KEY_MAX_PARAMETER_DESCRIPTION =
+      "[Future] The upper bound of the content key range to retrieve (exclusive).\n\n"
```
Since `maxKey` is inclusive, that wouldn't work.

dimas-b replied:
Yes. Having a separate `prefixKey` parameter is preferable :+1:

dimas-b commented on versioned/storage/store/src/test/java/org/projectnessie/versioned/storage/versionstore/TestKeyRanges.java:
```diff
@@ -0,0 +1,117 @@
+    soft.assertThat(keyRange.endStoreKey()).isGreaterThan(keyToStoreKeyVariant(bar, "A"));
+
+    soft.assertThat(keyRange.endStoreKey())
+        .isLessThanOrEqualTo(keyToStoreKey(ContentKey.of("bar", "baz")));
```
Why not simply `isLessThan`? We know it is not equal.

dimas-b commented on versioned/persist/store/src/main/java/org/projectnessie/versioned/persist/store/PersistVersionStore.java:
```diff
@@ -523,9 +524,18 @@ private BiConsumer<ImmutableCommit.Builder, CommitLogEntry> enhancerForCommitLog
-    checkArgument(pagingToken == null, "Paging not supported");
+    checkArgument(pagingToken == null, "Paging not supported by the storage model in use");
+    checkArgument(
+        minKey == null && maxKey == null, "Key ranges not supported by the storage model in use");
```
I guess we ought to check `prefixKey` too (also in `getDiffs`).

dimas-b commented on versioned/storage/store/src/test/java/org/projectnessie/versioned/storage/versionstore/TestKeyRanges.java:
```diff
@@ -0,0 +1,117 @@
+    soft.assertThat(keyRange.endStoreKey())
+        .isGreaterThan(keyToStoreKey(ContentKey.of("bar", "baz")));
+    soft.assertThat(keyRange.endStoreKey()).isGreaterThan(keyToStoreKey(ContentKey.of("bar", "0")));
+    soft.assertThat(keyRange.endStoreKey()).isGreaterThan(keyToStoreKey(ContentKey.of("barbaz")));
```
This fails, though:
```
    soft.assertThat(keyRange.endStoreKey()).isGreaterThan(keyToStoreKey(ContentKey.of("bar\u03C0")));
```

snazy replied:
Yes - why should it not? It's `> "bar"`

dimas-b replied:
But it has the right prefix.... why `barbaz` passes by `bar\u03C0` fails?

snazy replied:
It's a negative `byte` value in that UTF char - and this one just ends in a string comparison.

dimas-b replied:
but we support Unicode at the `ContentKey` level, right?.. and as a `ContentKey` is has the right prefix, so I guess it should be included in the range :thinking:

dimas-b replied:
To put it another way, should we use `Character.MAX_VALUE` instead of 255 in `TypeMapping`?

snazy replied:
Actually, it belongs to this PR. Just sad that this change requires such a change :(

dimas-b commented on versioned/storage/common/src/main/java/org/projectnessie/versioned/storage/common/indexes/StoreKey.java:
```diff
@@ -39,24 +41,54 @@ public final class StoreKey implements Comparable<StoreKey> {
+    if (diff == -1) {
+      return 0;
+    }
+    if (diff >= key.length) {
+      if (prefixEnd && diff == key.length) {
+        return 1;
+      }
+      return -1;
+    }
+    if (diff >= that.key.length) {
+      return 1;
+    }
```
This makes comparisons non-commutative. It will bite us sooner or later, I think.
```
  @Test
  void comparePrefix() {
    StoreKey k1 = StoreKey.keyFromStringPrefixEnd("aaa");
    StoreKey k2 = StoreKey.keyFromStringPrefixEnd("aaaBBB");
    soft.assertThat(k1).isGreaterThan(k2);
    soft.assertThat(k2).isLessThan(k1); // this assert fails
  }
```

dimas-b replied:
I think it might be preferable to actually encode the "prefix" flag in the key's byte form (to ensure consistency in case it ever gets into storage).

dimas-b commented on versioned/storage/common/src/main/java/org/projectnessie/versioned/storage/common/indexes/StoreKey.java:
```diff
@@ -39,24 +41,54 @@ public final class StoreKey implements Comparable<StoreKey> {
+    }
+    byte my = key[diff];
+    byte other = that.key[diff];
+    return Byte.compareUnsigned(my, other);
```
Is this always consistent with char-wise comparison?.. i.e. if we have two (unicode) strings, can we be sure that `StoreKeys` derived from those strings will compare the same way as the Strings?... more specifically - same as `ContentKey` objects?

snazy commented:
Reverted the `byte[]` change in `StoreKey` and updated `StoreKey.compareTo` to respect the `0`/`1` separated elements to not conflict with (unicode) chars.

dimas-b commented on versioned/storage/common/src/main/java/org/projectnessie/versioned/storage/common/indexes/StoreIndexImpl.java:
```diff
@@ -479,7 +479,23 @@ private ByteBuffer serializeKey(
+        // In PR #6743, introducing min/max/prefix-key arguments for "get entries" and "get diff",
+        // StoreKey.compareTo() was updated, which might lead to a different order in the elements
+        // list. The "current" version has been bumped, version=1 serialized indexes are sorted
+        // after they have been loaded.
```
Do you have an example where the sort order actually changes? I could not figure one out mentally :)

snazy replied:
ouch - again

dimas-b commented on versioned/storage/store/src/main/java/org/projectnessie/versioned/storage/versionstore/KeyRanges.java:
```diff
@@ -0,0 +1,88 @@
+  @Value.NonAttribute
+  StoreKey endStoreKey() {
+    ContentKey max = maxKey();
+    return max != null ? keyToStoreKeyMax(max) : null;
```
Why not derive a max key from `prefix` when it is set?

snazy replied:
That's not possible, sadly.
It would need something to compare to - but what?
"Incrementing the last char" doesn't really work there.

dimas-b commented on versioned/storage/store/src/main/java/org/projectnessie/versioned/storage/versionstore/VersionStoreImpl.java:
```diff
@@ -491,15 +502,15 @@ public PaginationIterator<KeyEntry> getKeys(Ref ref, String pagingToken, boolean
-
-    Iterator<StoreIndexElement<CommitOp>> result = index.iterator(begin, end, false);
+    Iterator<StoreIndexElement<CommitOp>> result =
+        index.iterator(keyRanges.beginStoreKey(), keyRanges.endStoreKey(), false);
```
If we do not derive an end key from `prefix`, then prefix queries will have to scan all the way to the end of the index, will they not?

snazy replied:
Fixed

dimas-b reviewed (commented):
Mostly optimization concerns

snazy commented on versioned/storage/common/src/main/java/org/projectnessie/versioned/storage/common/indexes/StoreKey.java:
```diff
@@ -51,7 +51,70 @@ public String rawString() {
+        return 1;
+      }
+
+      blkIdx1 = nextBlock(key, (char) 0, blkIdx1, blkLen1, blk1);
```
Um, totally right. Fixed.

dimas-b commented on versioned/storage/store/src/main/java/org/projectnessie/versioned/storage/versionstore/VersionStoreImpl.java:
```diff
@@ -748,16 +785,35 @@ public PaginationIterator<Diff> getDiffs(Ref from, Ref to, String pagingToken)
+    Predicate<DiffEntry> stopPredicate;
+    if (prefixKey != null) {
+      StoreKey prefix = keyToStoreKeyNoVariant(prefixKey);
+      stopPredicate = d -> !d.key().startsWithElementsOrParts(prefix);
```
`startsWithElementsOrParts` seems to contradict the doc text in `ApiDoc`... which way do we want it to work? :)

snazy replied:
doc's wrong, fixed

dimas-b reviewed (commented):
LGTM overall, but prefix handling seems to contradict (older) docs

snazy commented on api/model/src/main/java/org/projectnessie/api/v2/doc/ApiDoc.java:
```diff
@@ -109,12 +109,37 @@ public interface ApiDoc {
+  String KEY_PREFIX_PARAMETER_DESCRIPTION =
+      "The content key prefix to retrieve (inclusive). "
+          + "A content key matches a given prefix, a content key's elements starts with all elements of the prefix-key. "
+          + "It is not restricted to key-element boundaries.\n\n"
```
fixed
